gcs/gcs.go

```diff
@@ -74,16 +73,18 @@ func BuildGCSFilter(P uint8, key [KeySize]byte, data [][]byte) (*Filter, error)
 		p: P,
 	}
 	f.modulusP = uint64(1 << f.p)
-	f.modulusNP = uint64(f.n) * f.modulusP
 
 	// Build the filter.
 	values := make(uint64Slice, 0, len(data))
 	b := bstream.NewBStreamWriter(0)
 
 	// Insert the hash (modulo N*P) of each data element into a slice and
 	// sort the slice.
+	N := uint64(f.n)
 	for _, d := range data {
-		v := siphash.Sum64(d, &key) % f.modulusNP
+		v := siphash.Sum64(d, &key)
+		v = ((((v >> 32) * N) >> 32) << (32 - f.p)) |
```

Comment:
I think we should unpack this a bit, possible into distinct assignments with a brief comment explaining the rationale behind the shifts and OR's. As it exists now, I've grokked the computation, but it took a bit as everything is chained together. 

I'm assuming the compiler will eventually optimize it into a form like this, however for the sake of the reader I think we can make the operation a bit more clear.